### DISCUSSION TITLE
profile-controller doesn't work with 2 joysticks of same make, --next option removed?

### DISCUSSION THREAD

**Discussion by robincofer** [AntiMicroX/antimicrox#774] [Q&A] [answered]
I'm trying to set up a joystick-only interface on a gamebox (ubuntu 22.04), and am trying to automate some mappings with antimicrox 3.3.3. The best option would be if I could assign multiple profiles to multiple controllers in one command, but it appears this may no longer be supported.. ? When I use something like this:
`antimicrox --profile ${profile1_location} --profile-controller ${controller1_guid} --next --profile ${profile2_location} --profile-controller ${controller2_guid}`
I get the error "unknown option 'next'. If I remove the "--next" option from the above command, I get "WARN QCommandLineParser: option not defined: 'next'"
I see here the "--next" command line option seems to have been commented out:
https://github.com/AntiMicroX/antimicrox/blob/94a9523e582e5803b1d1d1f08b6143317d7790bc/src/commandlineutility.cpp#L103
The "--next" option is still listed in the man pages, but I'm guessing it's deprecated or something.. can anyone confirm that?

Regardless, I thought fine, I can just run a script with 2 separate commands. However, the GUID recognition seems to be broken.. I have 2 8bitdo controllers that are same model, with following output from the "--list" option (output edited for brevity):
Index:            1
  UniqueID:         050000005e040000e0020000030900001118736
  GUID:             050000005e040000e002000003090000
Index:            2
  UniqueID:         050000005e040000e0020000030900001118738
  GUID:             050000005e040000e002000003090000

When I run the command and set --profile-controller to the unique guid of one controller, it sets the profile for both controllers. It does not distinguish one from the other. 

I thought, ok, fine, I'll write a script to get the output of the list command, and get the index of the unique guid I'm trying to set, then run the command using index instead of guid. This results in the same thing, both controllers get set to the same profile, even though I use:
`antimicrox --profile /somefolder/someprofile.gamecontroller.amgp --profile-controller 1`
I tried indexes 1 and 2, and when that didn't work I thought maybe indexes were zero based and tried 0 and 1. Everything I try results in both controllers set to the same profile, whatever profile I set last.
Side note: the script I wrote to get the index from the guid was more complicated than it really should be, because `antimicrox --list` doesn't work if it's running.. you have to kill it first, run the list, then restart it. Would be a nice feature to have the list work while it's already running.

For various reasons, other features, such as loading profiles based on window properties, won't work for me. Can anyone confirm the above is a bug, or if I'm making some usage/syntax mistake? ..or anyone have alternate ideas about how to automate separate profiles on multiple controllers of same make?

**Comment by pktiuk** [comment, marked as answer]
`--next` is disabled but I don't know why. You can check: https://github.com/AntiMicroX/antimicrox/issues/189

**Reply by robincofer** [reply]
Thanks much for the reply.. should I submit the other issue as a bug? Seems to be:
"--profile-controller" option doesn't work when working with 2 controllers that have same guid and different unique ids

**Reply by pktiuk** [reply]
Firstly check whether similar issue already exists and only if you haven't sound anything similar open bug report.

**Reply by robincofer** [reply]
yes, had already looked, nothing specifically about this that I saw, tho whatever underlying issue may also be related to
https://github.com/AntiMicroX/antimicrox/issues/734
I'll do a little testing to isolate factors and submit an issue report specifically for the "--profile controller" option

**Reply by pktiuk** [reply]
Unluckily due to very limited time recently I don't know when I will be able to look deeper into this issue.

**Reply by pktiuk** [reply]
I just pushed a release with reenabled `--next` option, we'll see whether is useful. In the worst case, I will remove it.

**Comment by fsworld009** [comment]
Hi, I'm also trying to set different profiles to different controllers at once via command line and also facing the same issues.

While GUID related issues are tracked in #734, I'm wondering if we would re-enable `--next` option in a future release?
If it's not coming back, what would be the intended way to set different profiles to different controllers? (assuming all related bugs are fixed)

**Reply by pktiuk** [reply]
For now there are no plans for restoring this.     
I am not sure how that feature was supposed to work and why it was disabled. You can see: https://github.com/AntiMicroX/antimicrox/issues/189

**Comment by pmarks-net** [comment]
https://github.com/AntiMicroX/antimicrox/pull/1098 enabled the --next option, but the command-line parser was not written to handle repeated arguments based on position:

https://github.com/AntiMicroX/antimicrox/blob/50f3eb204cf1c40b1c31f92e09370118347a2c6e/src/commandlineutility.cpp#L112

Snippet:

```
    parser.process(parsed_app);
    int i = 0;
    while ((i < parser.optionNames().count()))
    {
        ...
        if (parser.isSet("next"))  // position ignored
        ...
        i++;  // i isn't used anywhere else
    }
```

**Reply by pmarks-net** [reply]
I think it's impossible to get this information from QCommandLineParser, so if `--profile --next --profile` is ever going to work, it needs an alternate (possibly custom) parser.

**Comment by pmarks-net** [comment]
I solved this problem by switching to https://github.com/sezanzeb/input-remapper:

```
input-remapper-control --command start --device "DragonRise Inc.   Generic   USB  Joystick  " --preset pinball-j1
input-remapper-control --command start --device "DragonRise Inc.   Generic   USB  Joystick   2" --preset pinball-j2
```
